File: 백준/Gold/9663. N－Queen/N－Queen.py
import sys


def n_queens(col, i):
    global cnt
    if promising(col, i):
        if i == n:
            cnt += 1
        else:
            for j in range(1, n + 1):
                col[i + 1] = j
                n_queens(col, i + 1)


def promising(col, i):
    k = 1
    flag = True
    while k < i and flag:
        if col[i] == col[k] or abs(col[i] - col[k]) == (i - k):
            flag = False
            break
        k += 1
    return flag


n = int(sys.stdin.readline())
cnt = 0
col = [0] * (n + 1)
n_queens(col, 0)
print(cnt)

File: 백준/Gold/9663. N－Queen/N－Queen.cc
#include <iostream>
#define MAX 15
using namespace std;

int n, cnt = 0;
int col[MAX] = { 0, };

bool promising(int i) {
	int k = 1;
	while (k < i) {
		if (col[i] == col[k] || abs(col[i] - col[k]) == i - k) {
			return false;
		}
		k++;
	}
	return true;
}

void nQueens(int i) {
	if (promising(i)) {
		if (i == n) cnt++;
		else {
			for (int j = 1; j <= n; j++) {
				col[i + 1] = j;
				nQueens(i + 1);
			}
		}
	}
}

int main() {
	cin.tie(0)->sync_with_stdio(0);
	
	cin >> n;
	nQueens(0);
	cout << cnt;
}